### PR TITLE
metrics: optimize CurrentFDs to not allocate on Linux

### PR DESCRIPTION
**Pull request by bradfitz**
It was 50% of our allocs on one of our servers. (!!)

Updates #2784

### CODE REVIEW

**Comment by bradfitz** [comment]
sorry

File: metrics/fds_linux.go [hunk 1]
```diff
@@ -4,10 +4,109 @@
...
+	nbuf := 0                      // end valid data in buf
+	dirent := &syscall.Dirent{}
+	var err error
+	for {
+		if bufp >= nbuf {
+			bufp = 0
+			nbuf, err = readDirent(fd, buf)
+			if err != nil {
+				log.Printf("currentFDs: readDirent: %v", err)
+				return 0
```
Comment (dsnet) [inline metrics/fds_linux.go:41]: In the event of an error, would returning the current count be better than just reporting zero?

File: metrics/fds_linux.go [hunk 2]
```diff
@@ -4,10 +4,109 @@
...
+		limit = nameBufLen
+	}
+	nameLen := bytes.IndexByte(nameBuf[:limit], 0)
+	if nameLen < 0 {
+		panic("failed to find terminating 0 byte in dirent")
+	}
+	return uint64(nameLen)
+}
+
+func parseDirEnt(dirent *syscall.Dirent, buf []byte) (consumed int, name []byte) {
```
Comment (crawshaw) [inline metrics/fds_linux.go:71]: Can this stuff go upstream in x/sys?

**Review by dsnet — APPROVED** [review]
Gross.

File: metrics/fds_linux.go (hunk 1 above)
Comment (bradfitz) [reply metrics/fds_linux.go:41]: I think we're more likely to notice if the graph drops to 0

File: metrics/fds_linux.go [hunk 3]
```diff
@@ -4,10 +4,109 @@
...
+	return
+}
+
+// According to https://golang.org/doc/go1.14#runtime
+// A consequence of the implementation of preemption is that on Unix systems, including Linux and macOS
+// systems, programs built with Go 1.14 will receive more signals than programs built with earlier releases.
+//
+// This causes syscall.Open and syscall.ReadDirent sometimes fail with EINTR errors.
+// We need to retry in this case.
+func open(path string, mode int, perm uint32) (fd int, err error) {
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply metrics/fds_linux.go:96]: whoops, fixed.

File: metrics/fds_linux.go (hunk 2 above)
Comment (bradfitz) [reply metrics/fds_linux.go:71]: It is, but it allocs a bunch there.

File: metrics/fds_linux.go [hunk 4]
```diff
@@ -4,10 +4,109 @@
...
+		if len(name) == 0 || string(name) == "." || string(name) == ".." {
+			continue
+		}
+		count++
+	}
+}
+
+func direntNamlen(dirent *syscall.Dirent) uint64 {
+	const fixedHdr = uint16(unsafe.Offsetof(syscall.Dirent{}.Name))
+	nameBuf := (*[unsafe.Sizeof(dirent.Name)]byte)(unsafe.Pointer(&dirent.Name[0]))
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply metrics/fds_linux.go:58]: This is code I had from elsewhere that I adapted. Simplified.

File: metrics/fds_linux.go (hunk 1 above)
Comment (dsnet) [reply metrics/fds_linux.go:41]: Fair argument.

**Comment by bradfitz** [comment]
PTAL

File: metrics/fds_linux.go (hunk 2 above)
Comment (crawshaw) [reply metrics/fds_linux.go:64]: I mean, can you upstream a new variant that doesn't alloc?

If we can't justify it upstream, can we justify carrying it here? Are these allocs important?

Comment (bradfitz) [reply metrics/fds_linux.go:64]: Started the process: https://github.com/golang/go/issues/48161

trunkd numbers says it's important

File: metrics/fds_linux.go [hunk 5]
```diff
@@ -4,10 +4,106 @@
...
+		consumed, name := parseDirEnt(dirent, buf[bufp:nbuf])
+		bufp += consumed
+		if len(name) == 0 || string(name) == "." || string(name) == ".." {
+			continue
+		}
+		count++
+	}
+}
+
+func direntNamlen(dirent *syscall.Dirent) uint64 {
```
Comment (crawshaw) [inline metrics/fds_linux.go:49]: Weird for a len of something held in memory to be a uint64 and not an int.

File: metrics/fds_linux.go [hunk 6]
```diff
@@ -4,10 +4,106 @@
...
+		if dirent.Name[i] == 0 {
+			return uint64(i)
+		}
+	}
+	panic("failed to find terminating 0 byte in dirent")
+}
+
+func parseDirEnt(dirent *syscall.Dirent, buf []byte) (consumed int, name []byte) {
+	// golang.org/issue/37269
+	copy((*[unsafe.Sizeof(syscall.Dirent{})]byte)(unsafe.Pointer(dirent))[:], buf)
```
Comment (crawshaw) [inline metrics/fds_linux.go:66]: unsafe.Slice and below?

**Review by crawshaw — APPROVED** [review]

File: metrics/fds_linux.go (hunk 6 above)
Comment (bradfitz) [reply metrics/fds_linux.go:66]: it's actually longer, but sure:

```
        copy((*[unsafe.Sizeof(syscall.Dirent{})]byte)(unsafe.Pointer(dirent))[:], buf)
=>
        copy(unsafe.Slice((*byte)(unsafe.Pointer(dirent)), unsafe.Sizeof(syscall.Dirent)), buf)
```

File: metrics/fds_linux.go (hunk 5 above)
Comment (bradfitz) [reply metrics/fds_linux.go:49]: done

File: metrics/fds_linux.go (hunk 6 above)
Comment (bradfitz) [reply metrics/fds_linux.go]: done

**Comment by evanphx** [comment]
Not to arm chair review a persons work, though I do have a question: Looking at unix.ParseDirent, could you have preallocated a large slice (`make([]string, 0, 10000)`) and passed that in as the `names` argument to unix.ParseDirent to prevent it from doing any allocations? Since you only care about the count variable, you don't even care about goroutines clobbering the contents of your names slice.

**Comment by dsnet** [comment]
> Not to arm chair review a persons work, though I do have a question: Looking at unix.ParseDirent, could you have preallocated a large slice (make([]string, 0, 10000)) and passed that in as the names argument to unix.ParseDirent to prevent it from doing any allocations? Since you only care about the count variable, you don't even care about goroutines clobbering the contents of your names slice.

Two thoughts:
* The `os` package lacks API to read into an existing `[]os.DirEntry`.
* I'm fairly certain the Go race detector won't be happen about concurrent mutations of the same slice, even if we don't care about the results. See https://github.com/golang/go/commit/eb43ce2d7711ad963de4860b70495a7aba3271c5 for a similar situation.

**Comment by bradfitz** [comment]
@evanphx, feel free to try it and see!

**Comment by evanphx** [comment]
@bradfitz: bitmoji

**Pull request merged**